### PR TITLE
Add target speaker extraction (TSE) functions

### PR DESCRIPTION
**Pull request by Emrys365**
This PR adds support for the target speaker extraction (TSE) task in addition to the existing speech enhancement/separation functions.

Major changes include:
1. Updated recipe template script in `egs2/TEMPLATE/enh1/enh.sh`
2. A TSE recipe in `egs2/librimix/tse1`
3. New TSE task in `espnet2/tasks/enh_tse.py`
   * New preprocessor `TSEPreprocessor` in `espnet2/train/preprocessor.py`
   * New training entry in `espnet2/enh/espnet_model_tse.py`
   * New inference script in `espnet2/bin/enh_tse_inference.py`
4. Implemented [time-domain SpeakerBeam (TD-SpeakerBeam)](https://arxiv.org/abs/2001.08378) model in `espnet2/enh/extractor/td_speakerbeam_extractor.py`

**Note:**

Current framework supports various ways of loading enrollments for each sample:
  1. fixed enrollment: explicitly specified audio/embedding path in `enroll_spk?.scp` in each subset
  2. dynamically sampled enrollments: following a special format in `enroll_spk?.scp` to inform the preprocessor to randomly sample the corresponding enrollment on the fly. See the definition of `TSEPreprocessor` in `espnet2/train/preprocessor.py` for more details.

  * Both audio and embeddings(in `npy` format) formats are supported. See the definition of `TSEPreprocessor` in `espnet2/train/preprocessor.py` for more details.

### CODE REVIEW

File: espnet2/train/preprocessor.py [hunk 1]
```diff
@@ -1175,3 +1177,131 @@ def __call__(
                 data[key] = data[key][:length]
 
         return data
+
+
+class TSEPreprocessor(AbsPreprocessor):
```
[the remark this answers is not in the document]
Comment (simpleoier) [reply espnet2/train/preprocessor.py:1182]: Could this be inheritated from EnhPreprocessor so that some data augmentation can still be applied?

**Review by simpleoier — COMMENTED** [review]
This is very cool new feature!

File: espnet2/train/preprocessor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Emrys365) [reply espnet2/train/preprocessor.py:1182]: Sure. Will do it.

File: egs2/TEMPLATE/enh1/tse.sh [hunk 2]
```diff
@@ -0,0 +1,1111 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (Emrys365) [reply egs2/TEMPLATE/enh1/tse.sh:1]: OK.

File: espnet2/enh/espnet_model_tse.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+from espnet2.enh.decoder.abs_decoder import AbsDecoder
+from espnet2.enh.encoder.abs_encoder import AbsEncoder
+from espnet2.enh.extractor.abs_extractor import AbsExtractor
+from espnet2.enh.loss.criterions.tf_domain import FrequencyDomainLoss
+from espnet2.enh.loss.criterions.time_domain import TimeDomainLoss
+from espnet2.enh.loss.wrappers.abs_wrapper import AbsLossWrapper
+from espnet2.torch_utils.device_funcs import force_gatherable
+from espnet2.train.abs_espnet_model import AbsESPnetModel
+
+is_torch_1_9_plus = V(torch.__version__) >= V("1.9.0")
```
[the remark this answers is not in the document]
Comment (Emrys365) [reply espnet2/enh/espnet_model_tse.py:17]: Oh I think this variable is useless here. Will remove it.

File: espnet2/enh/espnet_model_tse.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+            for spk in range(self.num_spk)
+            if "speech_ref{}".format(spk + 1) in kwargs
+        ]
+        # (Batch, num_speaker, samples) or (Batch, num_speaker, samples, channels)
+        speech_ref = torch.stack(speech_ref, dim=1)
+        batch_size = speech_mix.shape[0]
+
+        assert "enroll_ref1" in kwargs, "At least 1 enrollment signal is required."
+        # enrollment signal for each speaker (as the target)
+        enroll_ref = [
```
[the remark this answers is not in the document]
Comment (Emrys365) [reply espnet2/enh/espnet_model_tse.py:98]: Yes, the current design is one enrollment per speaker. For now, I cannot think of a good application where the two variables are not equal.

**Review by simpleoier — APPROVED** [review]
LGTM!

**Comment by Emrys365** [comment]
The CI error does not seem related to this PR.

<details><summary>Expand to see detailed errors</summary><div>

```
=================================== FAILURES ===================================
________ test_dynamic_quantization[quantize_dic0-recog_dic1-train_dic2] ________

train_dic = {'custom_enc_input_layer': 'linear', 'enc_block_arch': [{'dilation': 1, 'dropout-rate': 0.3, 'idim': 2, 'kernel_size': 2, ...}, {'d_ff': 2, 'd_hidden': 2, 'heads': 1, 'macaron_style': False, ...}]}
recog_dic = {'beam_size': 2, 'search_type': 'default'}
quantize_dic = {'dtype': torch.qint8, 'mod': {<class 'torch.nn.modules.linear.Linear'>}}

    @pytest.mark.parametrize(
        "train_dic",
        [
            {},
            {
                "enc_block_arch": [
                    {
                        "type": "conformer",
                        "d_hidden": 2,
                        "d_ff": 2,
                        "heads": 1,
                        "macaron_style": True,
                        "use_conv_mod": True,
                        "conv_mod_kernel": 1,
                    }
                ],
>           model.recognize(feats[0, : feats_len[0]].numpy(), beam_search)

test/test_custom_transducer.py:673: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
espnet/nets/pytorch_backend/e2e_asr_transducer.py:506: in recognize
    nbest_hyps = beam_search(enc_out)
espnet/nets/beam_search_transducer.py:151: in __call__
    nbest_hyps = self.search_algorithm(enc_out)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <espnet.nets.beam_search_transducer.BeamSearchTransducer object at 0x7f1ebaf43750>
enc_out = tensor([[ 1.0000, -1.0000],
        [ 1.0000, -1.0000],
        [ 1.0000, -1.0000],
        [ 1.0000, -1.0000],
      ...0000],
        [ 1.0000, -1.0000],
        [ 1.0000, -1.0000],
        [ 1.0000, -1.0000],
        [ 1.0000, -1.0000]])

    def default_beam_search(self, enc_out: torch.Tensor) -> List[Hypothesis]:
        """Beam search implementation.
    
        Modified from https://arxiv.org/pdf/1211.3711.pdf
    
        Args:
            enc_out: Encoder output sequence. (T, D)
    
        Returns:
            nbest_hyps: N-best hypothesis.
    
        """
        beam = min(self.beam_size, self.vocab_size)
        beam_k = min(beam, (self.vocab_size - 1))
    
        dec_state = self.decoder.init_state(1)
    
        kept_hyps = [Hypothesis(score=0.0, yseq=[self.blank_id], dec_state=dec_state)]
        cache = {}
    
        for enc_out_t in enc_out:
            hyps = kept_hyps
            kept_hyps = []
    
            while True:
                max_hyp = max(hyps, key=lambda x: x.score)
                hyps.remove(max_hyp)
    
                dec_out, state, lm_tokens = self.decoder.score(max_hyp, cache)
    
                logp = torch.log_softmax(
                    self.joint_network(
                        enc_out_t, dec_out, quantization=self.quantization
                    )
>                   / self.softmax_temperature,
                    dim=-1,
                )
E               Failed: Timeout >2.0s

espnet/nets/beam_search_transducer.py:287: Failed
```

</div></details>

File: espnet2/enh/espnet_model_tse.py [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+            speech_ref1: (Batch, samples)
+                        or (Batch, samples, channels)
+            speech_ref2: (Batch, samples)
+                        or (Batch, samples, channels)
+            ...
+            speech_mix_lengths: (Batch,), default None for chunk interator,
+                            because the chunk-iterator does not have the
+                            speech_lengths returned. see in
+                            espnet2/iterators/chunk_iter_factory.py
+            enroll_ref1: (Batch, samples_aux)
```
[the remark this answers is not in the document]
Comment (Emrys365) [reply espnet2/enh/espnet_model_tse.py:74]: Sure.

**Comment by sw005320** [comment]
@LiChenda, can you also review this PR?

File: egs2/librimix/tse1/conf/tuning/train_enh_tse_td_speakerbeam_4gpu_max.yaml [hunk 6]
```diff
@@ -0,0 +1,79 @@
+optim: adam
+max_epoch: 400
+batch_type: folded
+batch_size: 16 # batch_size 16 can be trained on 4 RTX 2080ti GPUs
```
Comment (LiChenda) [inline egs2/librimix/tse1/conf/tuning/train_enh_tse_td_speakerbeam_4gpu_max.yaml:4]: We'd better remove the comment.

File: egs2/librimix/tse1/local/data.sh [hunk 7]
```diff
@@ -0,0 +1,177 @@
+#!/usr/bin/env bash
```
Comment (LiChenda) [inline egs2/librimix/tse1/local/data.sh:1]: Is it possible to call some stage functions in `librimix/enh` recipe ? (Or just remove `data.sh` in `enh` recipe and just call this new one.) Maybe an unified `data.sh` for similar tasks in one dataset would be better.

File: espnet2/train/preprocessor.py [hunk 8]
```diff
@@ -198,7 +200,7 @@ def __init__(
                         self.noises.append(sps[1])
             sps = noise_db_range.split("_")
             if len(sps) == 1:
-                self.noise_db_low, self.noise_db_high = float(sps[0])
```
Comment (LiChenda) [inline espnet2/train/preprocessor.py:201]: looks like that the previous `self.noise_db_low, self.noise_db_high = float(sps[0])` is a bug?

**Review by LiChenda — APPROVED** [review]
LGTM! I just added some minor comments.

**Comment by sw005320** [comment]
@Emrys365, after you finish reflecting @LiChenda's comments, I'll merge this PR.

File: egs2/librimix/tse1/conf/tuning/train_enh_tse_td_speakerbeam_4gpu_max.yaml (hunk 6 above)
Comment (Emrys365) [reply egs2/librimix/tse1/conf/tuning/train_enh_tse_td_speakerbeam_4gpu_max.yaml:4]: Sure.

File: espnet2/train/preprocessor.py (hunk 8 above)
Comment (Emrys365) [reply espnet2/train/preprocessor.py:201]: Indeed. I believe this is a bug that was not found.

File: egs2/librimix/tse1/local/data.sh (hunk 7 above)
Comment (Emrys365) [reply egs2/librimix/tse1/local/data.sh:1]: I prefer maintaining `data.sh` for each sub-recipe, as they have different data requirements. For example, the TSE task requires additional enrollments as input.

> Is it possible to call some stage functions in `librimix/enh` recipe ?

Yes, I simply run through all stages in `librimix/enh1/local/data.sh` before the data preparation for TSE. However, the current `librimix/enh1/local/data.sh` script is not written in a flexible manner, and only supports a fixed number of speakers (num_spk=2) in the mixture. I just fix this part to support num_spk=3 as well.

**Comment by sw005320** [comment]
Thanks a lot, @Emrys365!
This is a great PR!

**Pull request merged**